File: community/cypher/src/main/scala/org/neo4j/cypher/internal/parser/v2_0/Labels.scala
/**
 * Copyright (c) 2002-2013 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher.internal.parser.v2_0

import org.neo4j.cypher.internal.commands.expressions.Expression
import org.neo4j.cypher.internal.commands.values.{LabelValue, LabelName}

trait Labels extends Base {
  def labelName: Parser[LabelName] = ":" ~> escapableString ^^ { x => LabelName(x) }

  def labelShortForm: Parser[Seq[LabelValue]] = rep1(labelName)

  def optLabelShortForm: Parser[Seq[LabelValue]] = opt(labelShortForm) ^^ {
    case optSpec => optSpec.getOrElse(Seq.empty)
  }

  def expression: Parser[Expression]
}

File: community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/Labels.java
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.neo4j.kernel.impl.newapi;

import java.util.Arrays;
import org.apache.commons.lang3.mutable.MutableInt;
import org.eclipse.collections.api.set.primitive.IntSet;
import org.eclipse.collections.api.set.primitive.LongSet;
import org.neo4j.internal.kernel.api.TokenSet;

public class Labels implements TokenSet {
    private final int[] labelIds;

    private Labels(int[] labelIds) {
        this.labelIds = labelIds;
    }

    public static Labels from(int... labels) {
        return new Labels(labels);
    }

    static Labels from(IntSet set) {
        return new Labels(set.toArray());
    }

    static Labels from(LongSet set) {
        final int[] labelIds = new int[set.size()];
        MutableInt index = new MutableInt();
        set.each(element -> labelIds[index.getAndIncrement()] = (int) element);
        return new Labels(labelIds);
    }

    @Override
    public int numberOfTokens() {
        return labelIds.length;
    }

    @Override
    public int token(int offset) {
        return labelIds[offset];
    }

    @Override
    public boolean contains(int token) {
        // It may look tempting to use binary search
        // however doing a linear search is actually faster for reasonable
        // label sizes (≤100 labels)
        for (int label : labelIds) {
            if (label == token) {
                return true;
            }
        }
        return false;
    }

    @Override
    public String toString() {
        return "Labels" + Arrays.toString(labelIds);
    }

    @Override
    public int[] all() {
        return labelIds;
    }
}
